shell/platform/android/io/flutter/plugin/common/JSONCodec.java

```diff
@@ -0,0 +1,79 @@
+package io.flutter.plugin.common;
+
+import java.nio.ByteBuffer;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+/**
+ * A {@link MessageCodec} using UTF-8 encoded JSON messages, and supporting the same Java values
+ * as {@link JSONObject#wrap(Object)}.
+ */
+final class JSONCodec implements MessageCodec {
+    @Override
+    public ByteBuffer encodeMessage(Object message) {
+        if (message == null) {
+            return null;
+        }
+        return ByteBufferTools.extractBytes(JSONObject.wrap(message).toString());
+    }
+
+    @Override
+    public Object decodeMessage(ByteBuffer message) {
+        if (message == null) {
+            return null;
+        }
+        try {
+            final String json = ByteBufferTools.extractString(message);
+            final JSONTokener tokener = new JSONTokener(json);
+            final Object value = tokener.nextValue();
+            if (tokener.more()) {
+                throw new IllegalArgumentException("Invalid JSON");
+            }
+            return value;
+        } catch (JSONException e) {
+            throw new IllegalArgumentException("Invalid JSON", e);
+        }
+    }
+
+    @Override
+    public MethodCall decodeMethodCall(ByteBuffer message) {
```

Comment:
It seems like this should be in a separate codec.  I would have expect the JSON codec to know about JSON but not about these particular JSON schema.